File: solution354.java


public class solution354 {
    public int maxEnvelopes(int[][] envelopes) {
        int maxW = 0, maxH = 0, res = 0;
        for (int i = 0; i < envelopes.length; i++) {
            if (envelopes[i][0] > maxW) {
                maxW = envelopes[i][0];
            }
            if (envelopes[i][1] > maxH) {
                maxH = envelopes[i][1];
            }
        }
        boolean[] usedW = new boolean[maxW], usedH = new boolean[maxH];
        for (int i = 0; i < envelopes.length; i++) {
            int w = envelopes[i][0], h = envelopes[i][1];
            if (!usedW[w - 1] && !usedH[h - 1]) {
                res++;
                usedW[w - 1] = true;
                usedH[h - 1] = true;
            }
        }
        return res;
    }
}

File: solution354.go
package leetcode

import (
	"sort"
)

func maxEnvelopes(envelopes [][]int) int {
	var n int = len(envelopes)
	if n < 1 {
		return 0
	}
	sort.Slice(envelopes, func(i, j int) bool {
		a, b := envelopes[i], envelopes[j]
		if a[0] < b[0] {
			return true
		} else if a[0] == b[0] {
			return a[1] >= b[1]
		}
		return false
	})
	dp := make([]int, n)
	dp[0] = 1
	res := 1
	for i := 1; i < n; i++ {
		maxLen := 0
		for j := i - 1; j >= maxLen; j-- {
			if envelopes[j][1] < envelopes[i][1] && dp[j] > maxLen {
				maxLen = dp[j]
			}
		}
		dp[i] = maxLen + 1
		if dp[i] > res {
			res = dp[i]
		}
	}
	return res
}
